### PR TITLE
Move from gRPC to HTTP

### PR DESCRIPTION
**Pull request by zltnDC**
TODO:
- [x] request timeouts should be the same as test timeouts
- [ ] Android - packagingOptions - is it possible to get rid of it ?
- [x] Android/example - throws PatrolAppServiceClientException - we should rather catch this error and not change the method.
- [x] grpc models had a pattern with hasProperty. What do we want to do with it ?
- [x] remove commented lines of code -> throw exception etc. - these lines helped in debugging corner cases

### CODE REVIEW

**Comment by shilangyu** [comment]
I would really like to hear the rationale of why a custom generator was chosen over something like openapi. Thread in #1532

**Comment by bartekpacia** [comment]
Issues to be closed after we release this:
- #828
- #1532
- #1660

File: .github/ISSUE_TEMPLATE/feature_request.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+name: Request a feature
```
Comment (bartekpacia) [inline .github/ISSUE_TEMPLATE/feature_request.yaml:1]: We should merge `master` into `develop` so this disappears.

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/Automator.kt [hunk 2]
```diff
@@ -334,29 +315,28 @@ class Automator private constructor() {
...
+                ?: notificationContainer.findObject(By.res("com.android.systemui:id/notification_title"))?.text
+            if (title == null) {
+                Logger.e("Could not find title text")
             }
+
+            val notification = Contracts.Notification(
+                appName = appName,
+                content = content ?: "",
+                title = title ?: "",
+                raw = "" //  Should it be optional ?
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/Automator.kt:337]: I think it's only set on iOS to workaround some limitation there. Hacky :(

Setting it to "Not available on Android" is good enough for now, I guess.

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/AutomatorServer.kt [hunk 3]
```diff
@@ -1,230 +1,207 @@
...
+    override fun setLocationAccuracy(request: Contracts.SetLocationAccuracyRequest) {
         when (request.locationAccuracy) {
-            COARSE -> automation.selectCoarseLocation()
-            FINE -> automation.selectFineLocation()
-            else -> throw PatrolException("setLocationAccuracy(): bad location accuracy")
+            Contracts.SetLocationAccuracyRequestLocationAccuracy.coarse
+            -> automation.selectCoarseLocation()
+
+            Contracts.SetLocationAccuracyRequestLocationAccuracy.fine
+            -> automation.selectFineLocation()
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/AutomatorServer.kt:182]: static imports, so the cases fit on a single lin

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/AutomatorServer.kt [hunk 4]
```diff
@@ -1,230 +1,207 @@
...
-            else -> throw PatrolException("tapOnNotification(): neither index nor selector are set")
+    override fun debug() {
+        // iOS only
+    }
+
+    override fun tapOnNotification(request: Contracts.TapOnNotificationRequest) {
+        if (request.index != null) {
+            automation.tapOnNotification(
+                request.index.toInt()
+            )
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/AutomatorServer.kt:194]: can be single line

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java [hunk 5]
```diff
@@ -18,7 +18,7 @@ public PatrolAppServiceClient createAppServiceClient() {
 
         try {
             client.listDartTests();
-        } catch (StatusRuntimeException ex) {
+        } catch (PatrolAppServiceClientException ex) {//TODO
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java:21]: why TODO here?

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java [hunk 6]
```diff
@@ -38,7 +38,7 @@ public String getLoopback() {
                 if (Objects.equals(i.getDisplayName(), "tun0")) {
                     for (java.net.InterfaceAddress a : i.getInterfaceAddresses()) {
                         if (a.getAddress() instanceof Inet4Address) {
-                            return a.getAddress().toString().substring(1) + ":8082";
+                            return a.getAddress().toString().substring(1);
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java:41]: why this change?

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/ContractsExtensions.kt [hunk 7]
```diff
@@ -95,44 +95,45 @@ fun Selector.toBySelector(): BySelector {
...
-        matchedFocused = true
-        By.focused(focused)
-    } else if (hasPkg()) {
-        matchedPkg = true
-        By.pkg(pkg)
-    } else {
-        throw IllegalArgumentException("SelectorQuery is empty")
-    }
+    var bySelector =
+        if (hasText()) {
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/ContractsExtensions.kt:99]: single line

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/ContractsExtensions.kt [hunk 8]
```diff
@@ -191,15 +192,21 @@ fun Selector.toBySelector(): BySelector {
...
 
-            tests.add(test.copy { name = "$parentGroupName ${test.name}" })
-        } else if (test.type == DartGroupEntry.GroupEntryType.GROUP) {
+            tests.add(
+                DartGroupEntry(
+                    name = "$parentGroupName ${test.name}",
+                    type = GroupEntryType.test,
+                    entries = listOf()
+                )
+            )
```
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/ContractsExtensions.kt:208]: does the code-generated `DartGroupEntry` not have `copyWith()`?

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolServerFilters.kt
Comment (bartekpacia) [inline packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolServerFilters.kt:1]: Did you consider using Ktor instead of http4k?

File: packages/patrol/android/src/test/kotlin/pl/leancode/patrol/ContractsExtensionsTest.kt [hunk 9]
```diff
@@ -2,51 +2,52 @@ package pl.leancode.patrol
...
-                dartTestGroup {
-                    name = "open_app_test"
-                    entries += listOf(dartTestCase { name = "open maps" })
-                },
-                dartTestGroup {
-                    name = "webview_test"
-                    entries += listOf(dartTestCase { name = "interacts with the LeanCode website in a webview" })
-                }
+        val dartTestGroup =
+            dartTestGroup(
```
Comment (bartekpacia) [inline packages/patrol/android/src/test/kotlin/pl/leancode/patrol/ContractsExtensionsTest.kt:24]: single line

File: packages/patrol/ios/Classes/AutomatorServer/Contracts.swift [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+enum SetLocationAccuracyRequestLocationAccuracy: String, Codable {
+  case coarse
+  case fine
+}
+
+struct DartGroupEntry: Codable {
+ var name: String
+ var type: GroupEntryType
+ var entries: [DartGroupEntry]
+}
```
Comment (bartekpacia) [inline packages/patrol/ios/Classes/AutomatorServer/Contracts.swift:39]: here - 1 space of indentation (not ok)

File: packages/patrol_gen/lib/src/generators/android/android_generator.dart [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+class AndroidGenerator {
+  List<OutputFile> generate(Schema schema, AndroidConfig config) {
+    final result = [
+      AndroidContractsGenerator().generate(schema, config),
+    ];
+
+    final serverGenerator = AndroidHttp4kServerGenerator();
+    final clientGenerator = AndroidHttp4kClientGenerator();
+
+    for (var service in schema.services) {
```
Comment (bartekpacia) [inline packages/patrol_gen/lib/src/generators/android/android_generator.dart:17]: `var -> final`

File: packages/patrol_gen/pubspec.yaml [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  sdk: '>=2.18.0 <4.0.0'
+
+dependencies:
+  meta: ^1.7.0
+  path: ^1.8.2
+  analyzer: ^6.1.0
+  dart_style: ^2.3.2
+  
+dev_dependencies:
+  leancode_lint: ^3.0.0
```
Comment (bartekpacia) [inline packages/patrol_gen/pubspec.yaml:19]: leancode_lint is added to pubspec but there's no analysis_options.yaml in the `patrol_gen` package. Is this intentional? Does the analysis work? (I suspect not because I noticed a few small things (e.g. unnecessary var) that leancode_lint would catch)

File: packages/patrol_gen/pubspec.yaml
Comment (bartekpacia) [inline packages/patrol_gen/pubspec.yaml:1]: `patrol_gen-prepare` Actions workflow that runs formatter and linter would be nice. You can copy&paste it from existing ones, they're almost all the same.

**Comment by bartekpacia** [comment]
I'm worried that `patrol_gen` has no tests.

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/Automator.kt (hunk 2 above)
Comment (zltnDC) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/Automator.kt:337]: In a private discussion, we decided to make this field optional and keep it consistent with the other fields.

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java (hunk 5 above)
Comment (zltnDC) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java:21]: I am not sure if it is this error or another exception from khttp will be correct in this case. So I asked you about a project where we use Browserstack, where we can verify this place :)

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java (hunk 6 above)
Comment (zltnDC) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java:41]: We pass the port to the client as a separate argument, we don't need to concatenate it here.

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolServer.kt [hunk 13]
```diff
@@ -2,44 +2,48 @@ package pl.leancode.patrol
...
     }
 
     private val arguments get() = InstrumentationRegistry.getArguments()
 
     fun start() {
+        Logger.i("Starting server...")
+
+        automatorServer = AutomatorServer(Automator.instance)
+        server =
+            automatorServer!!.router
```
[the remark this answers is not in the document]
Comment (zltnDC) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolServer.kt:29]: We always assign a fresh AutomatorServer instance one line of code above and don't refer to it other place. I don't think this is necessary.

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java (hunk 5 above)
Comment (bartekpacia) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/BrowserstackPatrolJUnitRunner.java:21]: OK, let's merge it as is into `develop` and I'll check if it works on BrowserStack.

File: packages/patrol/example/android/app/build.gradle [hunk 14]
```diff
@@ -32,6 +32,13 @@ android {
     kotlinOptions {
         jvmTarget = "1.8"
     }
+    
+    //TODO verify
+    packagingOptions {
+        pickFirst "META-INF/INDEX.LIST"
+        pickFirst "META-INF/io.netty.versions.properties"
+        pickFirst "META-INF/DEPENDENCIES"
+    }
```
[the remark this answers is not in the document]
Comment (zltnDC) [reply packages/patrol/example/android/app/build.gradle:41]: I spent some time to find another solution. This problem is similar to this one (link) and others related to "META/sth".
https://stackoverflow.com/questions/44342455/more-than-one-file-was-found-with-os-independent-path-meta-inf-license
There are different answers in this topic. I think this is a configuration issue with example app. It would be crazy if khttp dependencies in the same version caused metadata conflicts. We can delete the android project and configure it again. 
We can remove plugins or even change the server implementation, but I think it would be easier to try in other projects whether this problem occurs. If you want I can ask testers to use patrol with git commit and check this without merging and prerelease version.

File: packages/patrol/ios/Classes/ObjCPatrolAppServiceClient.swift [hunk 15]
```diff
@@ -1,9 +1,5 @@
-import GRPC
-import Logging
-import NIO
-
 /// The sole reason for existence of this class is that Swift Protobufs can't be used in Objective-C.
-@objc public class RunDartTestResponse: NSObject {
+@objc public class ObjCRunDartTestResponse: NSObject {
```
[the remark this answers is not in the document]
Comment (zltnDC) [reply packages/patrol/ios/Classes/ObjCPatrolAppServiceClient.swift]: I think we can leave this class, it's less complex.

File: schema.dart [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+enum SetLocationAccuracyRequestLocationAccuracy {
+  coarse,
+  fine,
+}
+
+class SetLocationAccuracyRequest {
+  late SetLocationAccuracyRequestLocationAccuracy locationAccuracy;
+}
+
+abstract class NativeAutomator<IOSServer, AndroidServer, DartClient> {
```
[the remark this answers is not in the document]
Comment (zltnDC) [reply schema.dart:159]: no. I will add simple documentation at the top of the file

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolServerFilters.kt
Comment (zltnDC) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolServerFilters.kt:1]: no, I used http4k because it worked before.

File: packages/patrol_gen/pubspec.yaml (hunk 12 above)
Comment (zltnDC) [reply packages/patrol_gen/pubspec.yaml:19]: I forgot to copy&paste

**Review by bartekpacia — APPROVED** [review]
please also merge with `master` to get CI to run. it stopped running after base branch was changed to `develop`. See #1696

File: schema.dart [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+  String? className;
+  String? contentDescription;
+  String? contentDescriptionStartsWith;
+  String? contentDescriptionContains;
+  String? resourceId;
+  int? instance;
+  bool? enabled;
+  bool? focused;
+  String? pkg;
+}
```
[the remark this answers is not in the document]
Comment (zltnDC) [reply schema.dart:65]: We can easily add this functionality later. I also think exporting internal models is not a best practice.
